File: observer-pattern/java/src/ConcreteObserver2.java
package src;

// 具体的观察者实现类，也可以看成订阅者，关联对应的主题类。
// 不同的观察者可以对应不同的主题。
public class ConcreteObserver2 implements ObserverAPI {

  // 这里没有在构造器就绑定某个主题，而是从客户角度去注册观察者
  public ConcreteObserver2() {
  }

  // 观察者发出更新通知，观察者自行监听
  public void update(String content) {
    System.out.println(String.format("%s::update() [content = %s]",
        this.getClass().getName(), content));
  }
}

File: observer-pattern/c/src/concrete_observer2.c
#include "func.h"

void concrete_observer2_set_subject(ConcreteObserver2 *observer, Subject *subject)
{
  observer->subject = subject;
}

// 观察者发出更新通知，不用单独告诉订阅者，由订阅者自行监听
void concrete_observer2_update(ConcreteObserver2 *observer, char *content)
{
  printf("\r\n ConcreteObserver2::update() [subject->name = %s content = %s]", observer->subject->name, content);
}

// 给观察者绑定主题，同时把观察者添加到主题列表
ConcreteObserver2 *concrete_observer2_constructor(char *name)
{
  printf("\r\n concrete_observer2_constructor() [构建ConcreteObserver2]");
  Observer *observer = (Observer *)malloc(sizeof(Observer));
  ConcreteObserver2 *concrete_observer2 = (ConcreteObserver2 *)observer;
  strncpy(observer->name, name, 50);
  concrete_observer2->set_subject = &concrete_observer2_set_subject;
  concrete_observer2->update = &concrete_observer2_update;

  return concrete_observer2;
}